src/mito2/src/manifest/manager.rs

```diff
@@ -327,68 +346,116 @@ impl RegionManifestManagerInner {
+        let last_checkpoint = Self::last_checkpoint(&self.store).await?;
+        let current_version = self.last_version;
+
+        let (start_version, mut manifest_builder) = if let Some(checkpoint) = last_checkpoint {
+            (
+                checkpoint.last_version + 1,
+                RegionManifestBuilder::with_checkpoint(checkpoint.checkpoint),
+            )
+        } else {
+            (MIN_VERSION, RegionManifestBuilder::default())
+        };
+        let end_version = current_version;
 
-    async fn scan(
-        &self,
-        start: ManifestVersion,
-        end: ManifestVersion,
-    ) -> Result<RegionMetaActionIter> {
-        todo!()
-    }
+        if start_version >= end_version {
+            return Ok(None);
+        }
 
-    async fn do_checkpoint(&self) -> Result<Option<RegionCheckpoint>> {
-        todo!()
-    }
+        let mut iter = self.store.scan(start_version, end_version).await?;
+        let mut last_version = start_version;
+        let mut compacted_actions = 0;
+        while let Some((version, raw_action_list)) = iter.next_log().await? {
+            let action_list = RegionMetaActionList::decode(&raw_action_list)?;
+            for action in action_list.actions {
+                match action {
+                    RegionMetaAction::Change(action) => {
+                        manifest_builder.apply_change(version, action);
+                    }
+                    RegionMetaAction::Edit(action) => {
+                        manifest_builder.apply_edit(version, action);
+                    }
+                    RegionMetaAction::Remove(_) | RegionMetaAction::Protocol(_) => {
+                        debug!(
+                            "Unhandled action for region {}, action: {:?}",
+                            self.manifest.metadata.region_id, action
+                        );
+                    }
+                }
+            }
+            last_version = version;
+            compacted_actions += 1;
+        }
 
-    async fn last_checkpoint(&self) -> Result<Option<RegionCheckpoint>> {
-        todo!()
-    }
+        if compacted_actions == 0 {
+            return Ok(None);
+        }
 
-    // from Checkpoint
+        let region_manifest = manifest_builder.try_build()?;
+        let checkpoint = RegionCheckpoint {
+            last_version,
+            compacted_actions,
+            checkpoint: Some(region_manifest),
+        };
 
-    /// Set a protocol action into checkpoint
-    pub fn set_protocol(&mut self, _action: ProtocolAction) {
-        todo!()
-    }
+        self.store
+            .save_checkpoint(last_version, &checkpoint.encode()?)
+            .await?;
+        // TODO(ruihang): this task can be detached
+        self.store.delete_until(last_version, true).await?;
 
-    /// The last compacted action's version of checkpoint
-    pub fn last_version(&self) -> ManifestVersion {
-        todo!()
+        info!(
```

Comment:
The log is annoying.

Reply:
You wrote this 🤨 https://github.com/GreptimeTeam/greptimedb/pull/1202/files#diff-77b390bae55c5b28b5f024a7fca31287eeea212ea623c84aa54e67b2b6365d01R130

Reply:
Downgrade it to `debug!`